src/BugsnagUnity/Delivery.cs

```diff
@@ -0,0 +1,172 @@
+                report.ApplyEventPayload();
+            }
+            SerializeAndDeliverPayload(payload);
+        }
+
+        IEnumerator PushToServer(IPayload payload, byte[] body)
+        {
+            using (var req = new UnityWebRequest(payload.Endpoint.ToString()))
+            {
+                req.SetRequestHeader("Content-Type", "application/json");
+                req.SetRequestHeader("Bugsnag-Sent-At", DateTimeOffset.Now.ToString("o", CultureInfo.InvariantCulture));
+                foreach (var header in payload.Headers)
+                {
+                    req.SetRequestHeader(header.Key, header.Value);
+                }
+                req.uploadHandler = new UploadHandlerRaw(body);
+                req.downloadHandler = new DownloadHandlerBuffer();
+                req.method = UnityWebRequest.kHttpVerbPOST;
+                yield return req.SendWebRequest();
+
+                while (!req.isDone)
+                {
+                    yield return new WaitForEndOfFrame();
+                }
+                var code = req.responseCode;
+                if (code == 200 || code == 202)
+                {
+                    // success!
+                    FileManager.PayloadSendSuccess(payload);
+                }
+                else if (code >= 500 || code == 408 || code == 429 || req.isNetworkError || code < 400)
+                {
+                    // sending failed, cache payload to disk
+                    FileManager.SendPayloadFailed(payload);
+                }
+                _finishedCacheDeliveries.Add(payload.Id);
+            }
+        }
+
+        public void TrySendingCachedPayloads()
+        {
+            if (_cacheDeliveryInProcess)
+            {
+                return;
+            }
+            _cacheDeliveryInProcess = true;
+            try
+            {
+                _finishedCacheDeliveries.Clear();
+                var payloads = FileManager.GetCachedPayloadPaths();
+                MainThreadDispatchBehaviour.Instance().Enqueue(DeliverCachedPayloads(payloads));
+            }
+            catch
+            {
+                // Not possible in unit tests
+            }
+        }
+
+        private bool CachedPayloadProcessed(string id)
+        {
+            foreach (var processedCachedPayloadIds in _finishedCacheDeliveries)
+            {
+                if (id == processedCachedPayloadIds)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private IEnumerator DeliverCachedPayloads(string[] payloadPaths)
+        {
+            foreach (var cachedPayloadPath in payloadPaths)
+            {
+                var payload = FileManager.GetPayloadFromCachePath(cachedPayloadPath);
+                if (payload != null)
+                {
+                    Send(payload);
+                    while (!CachedPayloadProcessed(payload.Id))
```

Comment:
Can you use [WaitUntil](https://docs.unity3d.com/ScriptReference/WaitUntil.html) checking the processed list for this rather than our own polling?

Reply:
nice, done 👍